search/includes/classes/class-versioning.php

```diff
@@ -316,4 +359,168 @@ public function activate_version( Indexable $indexable, int $version_number ) {
 	}
+
+	/**
+	 * Implements the vip_search_indexing_object_queued action to keep track of queued objects so that we can transparently
+	 * replicate the queued job to the non-active index versions
+	 * 
+	 *
+	 * @param int $object_id Object id
+	 * @param string $object_type Object type (the Indexable slug)
+	 * @param array $options The options passed to queue_object()
+	 * @param int $index_version The index version that was used when queuing the object
+	 */
+	public function action__vip_search_indexing_object_queued( $object_id, $object_type, $options, $index_version ) {
+		// Each time an object is queued, we keep track of that by version + type, so on shutdown, we can process them all in bulk
+		if ( ! isset( $this->queued_objects_by_type_and_version[ $object_type ] ) ) {
+			$this->queued_objects_by_type_and_version[ $object_type ] = array();
+		}
+
+		if ( ! isset( $this->queued_objects_by_type_and_version[ $object_type ][ $index_version ] ) ) {
+			$this->queued_objects_by_type_and_version[ $object_type ][ $index_version ] = array();
+		}
+
+		$this->queued_objects_by_type_and_version[ $object_type ][ $index_version ][] = array(
+			'object_id' => $object_id,
+			'options' => $options,
+		);
+	}
+
+	/**
+	 * When the request finishes, find all items that had been queued up on the active index and replicate those jobs out to each non-active index version
+	 * 
+	 * This ensures that the active index version is treated as The Truth, and non-active index versions follow it (and not the other way around)
+	 */
+	public function action__shutdown() {
+		$this->replicate_queued_objects_to_other_versions( $this->queued_objects_by_type_and_version );
+	}
+
+	/**
+	 * Given an array of object types and the objects queued by version, replicate those jobs to the
+	 * _other_ index versions to keep them in sync
+	 * 
+	 * @param $queued_objects Multidimensional array of queued objects, keyed first by type, then index version
+	 */
+	public function replicate_queued_objects_to_other_versions( $queued_objects ) {
+		if ( ! is_array( $queued_objects ) || empty( $queued_objects ) ) {
+			return;
+		}
+
+		// Loop over every type of object that was changed
+		foreach ( $queued_objects as $object_type => $objects_by_version ) {
+			$indexable = \ElasticPress\Indexables::factory()->get( $object_type );
+
+			// If it's not a valid indexable, just skip
+			if ( ! $indexable ) {
+				continue;
+			}
+
+			$versions = $this->get_versions( $indexable );
+
+			// Do we have any other index versions for this type? If not, nothing to do.
+			if ( ! $versions || empty( $versions ) ) {
+				continue;
+			}
+
+			$active_version_number = $this->get_active_version_number( $indexable );
+
+			// Were there any changes to the active version? If not, we skip - we don't keep replicate non-active indexes to others
+			if ( ! isset( $objects_by_version[ $active_version_number ] ) || empty( $objects_by_version[ $active_version_number ] ) ) {
+				continue;
+			}
```

Comment:
That's a good idea - the downside would be that you need to look up the active index version every time something is queued, which can sometimes be tens of thousands of times during a request. The `get_active_version_number()` is not _super_ expensive, but it could add up fast if called many times.

Reply:
Yep, you are right... I considered that, but I didn't take into account that this check is done much less often here. This was just about code clarity, but performance would be worse, so it's not worth changing anything